### PR TITLE
Fix rollback not leaving a consistent state for buffered transaction

### PR DESCRIPTION
**Pull request by domi4484**

### CODE REVIEW

**Comment by domi4484** [comment]
@nyalldawson related to #49639 this fixes the problem with the buffered transaction. The second problem is with the automatic transaction groups and I did not found the issue for now.

I have also changed a little the interface in `QgsVectorLayerEditBufferGroup`, a `SIP_OUT` was missing for some parameters. Do you think we could backport this API change to 3.26? This is really new and I don't think someone is already using that (`QgsVectorLayerEditBufferGroup` is also not intended to be used directly but via functions in `QgsProject`).

File: src/core/vector/qgsvectorlayereditbuffergroup.h [hunk 1]
```diff
@@ -91,14 +92,14 @@ class CORE_EXPORT QgsVectorLayerEditBufferGroup : public QObject
      *
      * \returns TRUE on success
      */
-    bool commitChanges( QStringList &commitErrors, bool stopEditing = true );
+    bool commitChanges( QStringList &commitErrors SIP_OUT, bool stopEditing = true );
```
Comment (nyalldawson) [inline src/core/vector/qgsvectorlayereditbuffergroup.h:95]: If you add "\param commitErrors will be set to a list of descriptive error messages if the commit fails" and "\param stopEditing .... etc" here, then sipify will properly handle SIP_OUT when generating Python docstrings. (same with rollback)

**Comment by nyalldawson** [comment]
Thanks @domi4484 !

> I have also changed a little the interface in QgsVectorLayerEditBufferGroup, a SIP_OUT was missing for some parameters. Do you think we could backport this API change to 3.26?

Yes, I agree -- it's a bug in the API

**Review by nyalldawson — APPROVED** [review]

**Comment by domi4484** [comment]
@nyalldawson regarding the second failing test, to me it seems an OGR problem. In "normal" transaction mode if an attribute is 
added to a layer it is not possible to remove it until the changes are committed. In the test output we see following:
```
ERROR 1: sqlite3_exec(ALTER TABLE "layer_a" DROP COLUMN "attr1") failed: no such column: ""attr1""
src/core/vector/qgsvectordataprovider.cpp:766 : (pushError) [63ms] OGR error deleting field 2: sqlite3_exec(ALTER TABLE "layer_a" DROP COLUMN "attr1") failed: no such column: ""attr1""
```
The test fails then at this [line](https://github.com/qgis/QGIS/blob/master/tests/src/python/test_qgsvectorlayereditbuffer.py#L668)


At the beginning i tout is an sqlite issue but if i try directly i can add and remove columns in an open transaction.
Used versions:
- gdal 3.4.2
- sqlite 3.39.2

**Pull request merged**